Flaky TestIngester_Push/should_discard_metadata_when_max_metadata_per_user_exceeded

From https://github.com/grafana/mimir/actions/runs/8838013308/job/24268182961?pr=7978

```
--- FAIL: TestIngester_Push (62.05s)
    --- FAIL: TestIngester_Push/should_discard_metadata_when_max_metadata_per_user_exceeded (2.00s)
        ingester_test.go:2614: 
            	Error Trace:	/__w/mimir/mimir/pkg/ingester/ingester_test.go:2614
            	Error:      	elements differ
            	            	
            	            	extra elements in list B:
            	            	([]interface {}) (len=1) {
            	            	 (*mimirpb.MetricMetadata)({
            	            	  Type: (mimirpb.MetricMetadata_MetricType) 1,
            	            	  MetricFamilyName: (string) (len=13) "test_metric_2",
            	            	  Help: (string) (len=22) "This is a test metric.",
            	            	  Unit: (string) ""
            	            	 })
            	            	}
            	            	
            	            	
            	            	listA:
            	            	([]*mimirpb.MetricMetadata) (len=1) {
            	            	 (*mimirpb.MetricMetadata)({
            	            	  Type: (mimirpb.MetricMetadata_MetricType) 1,
            	            	  MetricFamilyName: (string) (len=13) "test_metric_1",
            	            	  Help: (string) (len=22) "This is a test metric.",
            	            	  Unit: (string) ""
            	            	 })
            	            	}
            	            	
            	            	
            	            	listB:
            	            	([]*mimirpb.MetricMetadata) (len=2) {
            	            	 (*mimirpb.MetricMetadata)({
            	            	  Type: (mimirpb.MetricMetadata_MetricType) 1,
            	            	  MetricFamilyName: (string) (len=13) "test_metric_1",
            	            	  Help: (string) (len=22) "This is a test metric.",
            	            	  Unit: (string) ""
            	            	 }),
            	            	 (*mimirpb.MetricMetadata)({
            	            	  Type: (mimirpb.MetricMetadata_MetricType) 1,
            	            	  MetricFamilyName: (string) (len=13) "test_metric_2",
            	            	  Help: (string) (len=22) "This is a test metric.",
            	            	  Unit: (string) ""
            	            	 })
            	            	}
            	Test:       	TestIngester_Push/should_discard_metadata_when_max_metadata_per_user_exceeded
        ingester_test.go:2626: 
            	Error Trace:	/__w/mimir/mimir/pkg/ingester/ingester_test.go:2626
            	Error:      	Received unexpected error:
            	            	
            	            	
            	            	Diff:
            	            	--- metric output does not match expectation; want
            	            	+++ got:
            	            	@@ -1,16 +1,13 @@
            	            	-(bytes.Buffer) # HELP cortex_discarded_metadata_total The total number of metadata that were discarded.
            	            	-# TYPE cortex_discarded_metadata_total counter
            	            	-cortex_discarded_metadata_total{reason="per_user_metadata_limit",user="test"} 1
            	            	-# HELP cortex_ingester_ingested_metadata_failures_total The total number of metadata that errored on ingestion.
            	            	+(bytes.Buffer) # HELP cortex_ingester_ingested_metadata_failures_total The total number of metadata that errored on ingestion.
            	            	 # TYPE cortex_ingester_ingested_metadata_failures_total counter
            	            	-cortex_ingester_ingested_metadata_failures_total 1
            	            	+cortex_ingester_ingested_metadata_failures_total 0
            	            	 # HELP cortex_ingester_ingested_metadata_total The total number of metadata ingested.
            	            	 # TYPE cortex_ingester_ingested_metadata_total counter
            	            	-cortex_ingester_ingested_metadata_total 1
            	            	+cortex_ingester_ingested_metadata_total 2
            	            	 # HELP cortex_ingester_memory_metadata The current number of metadata in memory.
            	            	 # TYPE cortex_ingester_memory_metadata gauge
            	            	-cortex_ingester_memory_metadata 1
            	            	+cortex_ingester_memory_metadata 2
            	            	 # HELP cortex_ingester_memory_metadata_created_total The total number of metadata that were created per user
            	            	 # TYPE cortex_ingester_memory_metadata_created_total counter
            	            	-cortex_ingester_memory_metadata_created_total{user="test"} 1
            	            	+cortex_ingester_memory_metadata_created_total{user="test"} 2
            	            	 # HELP cortex_ingester_memory_series The current number of series in memory.
            	Test:       	TestIngester_Push/should_discard_metadata_when_max_metadata_per_user_exceeded
level=info msg="uploading new block to long-term storage" block=00000000010000000000000000
level=debug msg="uploaded file" from=/tmp/TestShipper_DeceivingUploadErrors386048170/001/00000000010000000000000000/index dst=00000000010000000000000000/index bucket="fs: /tmp/TestShipper_DeceivingUploadErrors386048170/002"
level=error msg="uploading new block to long-term storage failed" block=00000000010000000000000000 err="upload meta file: base name matches, will fail upload"
FAIL
FAIL	github.com/grafana/mimir/pkg/ingester	539.707s
```

--- a/pkg/ingester/ingester_test.go
+++ b/pkg/ingester/ingester_test.go
@@ -2549,9 +2549,10 @@ func TestIngester_Push(t *testing.T) {
 
 			ctx := user.InjectOrgID(context.Background(), userID)
 
-			// Wait until the ingester is healthy
-			test.Poll(t, 100*time.Millisecond, 1, func() interface{} {
-				return i.lifecycler.HealthyInstancesCount()
+			// Wait until the ingester is healthy and owns tokens. Note that the timeout here is set
+			// such that it is longer than the MinReadyDuration configuration for the ingester ring.
+			test.Poll(t, time.Second, nil, func() interface{} {
+				return i.lifecycler.CheckReady(context.Background())
 			})
 
 			// Push timeseries

--- a/pkg/ingester/lifecycle_test.go
+++ b/pkg/ingester/lifecycle_test.go
@@ -45,6 +45,7 @@ func defaultIngesterTestConfig(t testing.TB) Config {
 	cfg.IngesterRing.InstanceAddr = "localhost"
 	cfg.IngesterRing.InstanceID = "localhost"
 	cfg.IngesterRing.FinalSleep = 0
+	cfg.IngesterRing.MinReadyDuration = 100 * time.Millisecond
 	cfg.ActiveSeriesMetrics.Enabled = true
 
 	return cfg

